### PR TITLE
fix: XMLHttp Request Error

### LINKED ISSUE #641: XMLHttp Request Error
**Is your feature request related to a problem? Please describe.**
For the flutter web mentorship client, i keep getting a XMLHTTP Error. This is the exact error.
``` Access to XMLHttpRequest at 'https://mentorship-backend-temp.herokuapp.com/login' from origin '<myurl>' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.```

**Describe the solution you'd like**
 A solution for this would be enabling CORS. This should be pretty easy. You may refer [this]( https://flask-cors.readthedocs.io/en/latest/)

### PR DESCRIPTION
**Pull request by satya7289**
### Description

Fix-> XMLHttp Request Error by adding the Flask-CORS package for all routes in the application.

Fixes #641

### Type of Change:

- Code
- Quality Assurance
- Documentation

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)

### How Has This Been Tested?

Locally by AJAX request.

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] Update requirements.txt

**Code/Quality Assurance Only**

- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by Techno-Disaster** [comment]
Don't we also need the ``` cors.init_app(app, resources={r"*": {"origins": "http://localhost:3000"}}) ``` line?

**Comment by techno-disaster** [comment]
@mtreacy002 Can you give this a quick look? Is this how you did it on the bridge-in-tech backend?

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@satya7289 and @Techno-Disaster , Shouldn't you also declare the resources parameters to specify which type of resources and origin the request coming from?

I think not specifying resources like you did here will increase security risk.

You can read more [here](https://flask-cors.readthedocs.io/en/latest/). I did the `resources specific CORS` in my bridge-in-tech-backend

File: run.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
 from flask import Flask
 from config import get_env_config
 from flask_migrate import Migrate
+from flask_cors import CORS
 
 
 def create_app(config_filename: str) -> Flask:
     app = Flask(__name__)
 
+    # setup for CORS policy
```
Comment (mtreacy002) [inline run.py:10]: Here's the [zulip link](https://anitab-org.zulipchat.com/#narrow/stream/237630-bridge-in-tech/topic/Troubleshoot/near/200494292) where I mentioned the steps I've done on bridge-in-tech-backend 😉

**Comment by satya7289** [comment]
> @satya7289 and @Techno-Disaster , Shouldn't you also declare the resources parameters to specify which type of resources and origin the request coming from?
> 
> I think not specifying resources like you did here will increase security risk.
> 
> You can read more [here](https://flask-cors.readthedocs.io/en/latest/). I did the `resources specific CORS` in my bridge-in-tech-backend

You are right but please let me know that `Flutter` has any host server(like in your case react uses http://localhost:3000) so that it can be restricted to those host origin only.
@mtreacy002 @Techno-Disaster

**Comment by techno-disaster** [comment]
@satya7289 for now its hosted here  https://musing-mirzakhani-5ab76a.netlify.app/#/  add this for now, we will update the link later

**Comment by satya7289** [comment]
> @satya7289 for now its hosted here https://musing-mirzakhani-5ab76a.netlify.app/#/ add this for now, we will update the link later

But for the developing locally `Flutter` has no hosting server then these changes will be fine. Like you start the localhost server for mentorship backend and requesting the server from the `Flutter` app will work here. I think no need to specify the origin here. @mtreacy002

**Comment by mtreacy002** [comment]
> > @satya7289 for now its hosted here https://musing-mirzakhani-5ab76a.netlify.app/#/ add this for now, we will update the link later
> 
> But for the developing locally `Flutter` has no hosting server then these changes will be fine. Like you start the localhost server for mentorship backend and requesting the server from the `Flutter` app will work here. I think no need to specify the origin here. @mtreacy002

@satya7289, the issue will come sooner than you think in the case of Mentorship, I believe, since it is nearing to production (in fact there is a team designated for it to be "production-ready" asap). With this in mind, it is best to deal with it now, in my opinion. cc @Techno-Disaster and @isabelcosta.
 
If we leave it as it is now, when Mentorship System goes to production, there's a risk that this url gets overlooked which will "open a window" to attackers (coz send request from flutter will still work as normal). Whereas if we specify it as localhost, when it ever gets carried to production (which it shouldn't - if we have staging pipeline to check app readiness), It will make the app  failed to perform send request from Flutter, which is good, meaning someone will notice the bug and the localhost url needs to be replaced with a proper one.

**Comment by satya7289** [comment]
> > > @satya7289 for now its hosted here https://musing-mirzakhani-5ab76a.netlify.app/#/ add this for now, we will update the link later
> > 
> > 
> > But for the developing locally `Flutter` has no hosting server then these changes will be fine. Like you start the localhost server for mentorship backend and requesting the server from the `Flutter` app will work here. I think no need to specify the origin here. @mtreacy002
> 
> @satya7289, the issue will come sooner than you think in the case of Mentorship, I believe, since it is nearing to production (in fact there is a team designated for it to be "production-ready" asap). With this in mind, it is best to deal with it now, in my opinion. cc @Techno-Disaster and @isabelcosta.
> 
> If we leave it as it is now, when Mentorship System goes to production, there's a risk that this url gets overlooked which will "open a window" to attackers (coz send request from flutter will still work as normal). Whereas if we specify it as localhost, when it ever gets carried to production (which it shouldn't - if we have staging pipeline to check app readiness), It will make the app failed to perform send request from Flutter, which is good, meaning someone will notice the bug and the localhost url needs to be replaced with a proper one.

So, what you suggest here. @mtreacy002 
One thing I missed that he is not requesting from localhost actually he is requesting from `https://mentorship-backend-temp.herokuapp.com`

**Comment by mtreacy002** [comment]
@satya7289 , I'm just referring to your post here on running local server for mentorship backend
> But for the developing locally Flutter has no hosting server then these changes will be fine. Like you start the localhost server for mentorship backend and requesting the server from the Flutter app will work here. I think no need to specify the origin here. @mtreacy002

If @Techno-Disaster is going to be requesting to mentorship-heroku server, as per your point here
> One thing I missed that he is not requesting from localhost actually he is requesting from https://mentorship-backend-temp.herokuapp.com

then the need for stating the origin url of Flutter on the Mentorship code base will be even more crucial, because Heroku is not a local environment which means we can't just accept any requests especially from unapproved origin.

**Comment by mtreacy002** [comment]
What I want to say is, IMO it doesn't hurt to take additional measures in security related matters. I understand it's impossible to be totally free of risk (hackers can always find their way to break into the system), regardless, if we can add extra layers in place, it's better than relaxing it all together, right?

**Comment by techno-disaster** [comment]
> @satya7289 , I'm just referring to your post here on running local server for mentorship backend
> 
> > But for the developing locally Flutter has no hosting server then these changes will be fine. Like you start the localhost server for mentorship backend and requesting the server from the Flutter app will work here. I think no need to specify the origin here. @mtreacy002
> 
> If @Techno-Disaster is going to be requesting to mentorship-heroku server, as per your point here
> 
> > One thing I missed that he is not requesting from localhost actually he is requesting from https://mentorship-backend-temp.herokuapp.com
> 
> then the need for stating the origin url of Flutter on the Mentorship code base will be even more crucial, because Heroku is not a local environment which means we can't just accept any requests especially from unapproved origin.

yes, i will be making request directly to the heroku server, hence i sent the netifly link above. I have hosted the flutter app there temporarily, we will change it once i show it in the project roadmap session and get reviews about it from the admins. @satya7289 @mtreacy002

**Comment by adi-001** [comment]
> What I want to say is, IMO it doesn't hurt to take additional measures in security related matters. I understand it's impossible to be totally free of risk (hackers can always find their way to break into the system), regardless, if we can add extra layers in place, it's better than relaxing it all together, right?

Hi @mtreacy002 , i totally get your concern here so what alternate ways do you suggest ?
I googled stuffs online & got a [nice blog](https://medium.com/@dtkatz/3-ways-to-fix-the-cors-error-and-how-access-control-allow-origin-works-d97d55946d9). I personally suggest the third way , what do think about it?

File: run.py (hunk 1 above)
Comment (adi-001) [reply run.py:10]:
```suggestion
    # setup for CORS policy
```
i too believe this way it's likely to face a problem in near future

**Review by adi-001 — COMMENTED** [review]
At present, the issue might get resolved but believe  open to attacks later when placed in prod-server

**Comment by mtreacy002** [comment]
> > What I want to say is, IMO it doesn't hurt to take additional measures in security related matters. I understand it's impossible to be totally free of risk (hackers can always find their way to break into the system), regardless, if we can add extra layers in place, it's better than relaxing it all together, right?
> 
> Hi @mtreacy002 , i totally get your concern here so what alternate ways do you suggest ?
> I googled stuffs online & got a [nice blog](https://medium.com/@dtkatz/3-ways-to-fix-the-cors-error-and-how-access-control-allow-origin-works-d97d55946d9). I personally suggest the third way , what do think about it?

@adi-001 . The third one is a nice work around. we just need to make sure the same solution can be applied on python flaskrestx/flask-cors 😉. 
Although, the general consensus towards `*` wildcard on Allow-Control-Allow-Origin seems to the same, which is that it is highly risky (found these article [here](https://portswigger.net/web-security/cors/access-control-allow-origin) and [here](https://www.pivotpointsecurity.com/blog/cross-origin-resource-sharing-security/) that have same vote). So, regardless the workaround, if it still uses the wildcard "*", I think we need to avoid it. This is just my opinion.

**Comment by techno-disaster** [comment]
@satya7289 @isabelcosta Is there anything blocking this PR?

**Comment by epicadk** [comment]
Closing as issue linked was closed

**Pull request closed without merging**